File: Linked-List/doubly-linked-list.js
class Node {
    constructor(value) {
        this.value = value;
        this.next = null;
        this.prev = null;
    }
}


class DoublyLinkedList {
    constructor(value) {
        this.head = new Node(value);
        this.tail = this.head;
        this.length = 1;
    }
    append(value) {
        const newNode = new Node(value);
        newNode.prev = this.tail;
        this.tail.next = newNode;
        this.tail = newNode;
        this.length++;
        // console.log(this);
        return this;
    }
    prepend(value) {
        const newNode = new Node(value);
         newNode.next = this.head;
         this.head = newNode;
         this.length++;
         return this;
    }
    printList() {
        const arr = [];
        let currentNode = this.head;
        while(currentNode != null) {
            arr.push(currentNode.value);
            currentNode = currentNode.next;
        }
        console.log(arr);
        return arr;
    }
    insert(index, value) {
        const newNode = new Node(value);
        const leader = this.traverseToIndex(index-1);
        const follower = leader.next;
        leader.next = newNode;
        newNode.prev = leader;
        newNode.next = follower;
        follower.prev = newNode;
        this.length++;
        return this.printList();
    }

    remove(index) {
        const leader = this.traverseToIndex(index-1);
        const unwantedNode = leader.next;
        leader.next = unwantedNode.next;
        this.length--;
        return this;
    }
    traverseToIndex(index) {
        let counter = 0;
        let currentNode = this.head;
        while(counter !== index) {
            currentNode = currentNode.next;
            counter++;
        }
        return currentNode;
    }
    reverse() {
        console.log('Before.......');
        console.log(this);
        if (!this.head.next) {
            return this.head;
        }
        let first = this.head;
        this.tail = this.head;
        let second = first.next;
        while(second) {
            // console.log(second);
            const third = second.next;
            second.next = first;
            first = second;
            second = third;
        }
        this.head.next = null;
        this.head = first;
        console.log('After.......');
        console.log(this);
        return this;
    }

    

}

const myLinkedList = new DoublyLinkedList(1);
myLinkedList.append(2);
myLinkedList.append(3);
// myLinkedList.append(17);
// myLinkedList.insert(2, 100);
// myLinkedList.remove(1);
// myLinkedList.prepend(1);
// myLinkedList.printList();
// console.log(myLinkedList);
myLinkedList.reverse();

File: JAVA_VERSION/DATA-STRUCTURE_ALGORITHM/src/datastructure/linkedlist/DoublyLinkedList.java
package datastructure.linkedlist;


public class DoublyLinkedList {
    private int length;
    Node head;
    Node tail;
    class Node {
        private int value;
        private Node next;
        private Node prev;

        Node(int value) {
            this.value = value;
        }

        public int getValue() {
            return value;
        }

        public void setValue(int value) {
            this.value = value;
        }

        public Node getNext() {
            return next;
        }

        public void setNext(Node next) {
            this.next = next;
        }

        public Node getPrev() {
            return prev;
        }

        public void setPrev(Node prev) {
            this.prev = prev;
        }
    }
    DoublyLinkedList(int value) {
        Node newNode = new Node(value);
        this.head = newNode;
        this.tail = newNode;
        this.length++;
    }
    public void printList() {
        Node temp = this.head;
        while(temp!= null) {
            System.out.println(temp.value);
            temp = temp.next;
        }
    }
    public void prepend(int value) {
        Node newNode = new Node(value);

        if(this.length == 0) {
            this.head = newNode;
            this.tail = newNode;
        } else {
            newNode.setNext(this.head);
            this.head = newNode;
        }
        this.length++;
    }

    public void append(int value) {
        Node newNode = new Node(value);
        if(this.length == 0) {
            this.head = newNode;
            this.tail = newNode;
        } else {
            this.tail.next = newNode;
            newNode.prev = this.tail;
            this.tail = newNode;
            this.length++;
        }

    }
    public boolean insert(int index, int value) {
        Node newNode = new Node(value);
        if (index < 0 || index > length) return false;
        if (index == 0) {
            prepend(value);
            return true;
        }
        if(index == length) {
            append(value);
            return true;
        }
        Node current = get(index);
        Node prev = current.getPrev();
        newNode.setNext(current);
        newNode.setPrev(prev);
        current.setPrev(newNode);
        prev.setNext(newNode);
        this.length++;
        return true;
    }
    public Node remove(int index) {
        Node current = get(index);

        if(index == 0) {
            return this.removeFirst();
        }
        if(index == this.length-1) {
            return this.removeLast();
        }
        current.getNext().getPrev().setPrev(current.getPrev());
        current.getPrev().setNext(current.getNext());
        current.setNext(null);
        current.setPrev(null);
        this.length--;
        return current;
    }
    public Node get(int index) {
        Node temp = this.head;
        if (index < 0 || index > length) return null;

        if(index < length/2) {
            for (int i=0; i < index; i++) {
                temp = temp.getNext();
            }
        } else {
            temp = this.tail;
            for(int i = length-1; i > index; i--) {
                temp = temp.getPrev();
            }
        }
        return temp;
    }
    public boolean set(int index, int value) {
        Node temp = get(index);
        if (temp != null) {
            temp.value = value;
            return true;
        }
        return false;
    }

    public Node removeFirst() {
        Node temp = this.head;
        if(this.length == 0) return null;
        if(this.length == 1) {
            this.head = null;
            this.tail = null;
        } else {
            this.head = this.head.getNext();
            this.head.prev = null;
            temp = null;
            this.length--;
        }
        return  temp;
    }
    public Node removeLast() {
        Node temp = this.tail;
        if(this.length == 0) return null;
        if(this.length == 1) {
            this.head = null;
            this.tail = null;
        } else {
            this.tail = this.tail.prev;
            this.tail.setNext(null);
        }

        this.length--;

        return temp;

    }
    public void getHead() {
        System.out.println("Head: " + this.head.getValue());
    }

    public void getTail() {
        System.out.println("Tail: " + this.tail.getValue());
    }

    public void getLength() {
        System.out.println("Length: " + this.length);
    }
}
